opentelemetry-sdk/src/logs/growable_array.rs

```diff
@@ -0,0 +1,323 @@
+use std::array::IntoIter;
+
+/// The default max capacity for the stack portion of `GrowableArray`.
+const DEFAULT_MAX_STACK_CAPACITY: usize = 10;
+/// The default initial capacity for the vector portion of `GrowableArray`.
+const DEFAULT_INITIAL_VEC_CAPACITY: usize = 5;
+
+#[derive(Debug, Clone, PartialEq)]
+/// A hybrid vector that starts with a fixed-size array and grows dynamically with a vector.
+///
+/// `GrowableArray` uses an internal fixed-size array (`initial`) for storing elements until it reaches
+/// `MAX_STACK_CAPACITY`. When this capacity is exceeded, additional elements are stored in a heap-allocated
+/// vector (`additional`). This structure allows for efficient use of stack memory for small numbers of elements,
+/// while still supporting dynamic growth.
+///
+pub(crate) struct GrowableArray<
+    T: Default + Clone + PartialEq,
+    const MAX_STACK_CAPACITY: usize = DEFAULT_MAX_STACK_CAPACITY,
+    const INITIAL_VEC_CAPACITY: usize = DEFAULT_INITIAL_VEC_CAPACITY,
+> {
+    initial: [T; MAX_STACK_CAPACITY],
+    additional: Option<Vec<T>>,
+    count: usize,
+}
+
+impl<
+        T: Default + Clone + PartialEq,
+        const MAX_STACK_CAPACITY: usize,
+        const INITIAL_VEC_CAPACITY: usize,
+    > Default for GrowableArray<T, MAX_STACK_CAPACITY, INITIAL_VEC_CAPACITY>
+{
+    fn default() -> Self {
+        Self {
+            initial: [(); MAX_STACK_CAPACITY].map(|_| T::default()),
+            additional: None,
+            count: 0,
+        }
+    }
+}
+
+impl<
+        T: Default + Clone + PartialEq,
+        const MAX_STACK_CAPACITY: usize,
+        const INITIAL_VEC_CAPACITY: usize,
+    > GrowableArray<T, MAX_STACK_CAPACITY, INITIAL_VEC_CAPACITY>
+{
+    /// Creates a new `GrowableArray` with the default initial capacity.
+    #[allow(dead_code)]
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    /// Pushes a value into the `GrowableArray`.
+    ///
+    /// If the internal array (`initial`) has reached its capacity (`MAX_STACK_CAPACITY`), the value is pushed
+    /// into the heap-allocated vector (`additional`). Otherwise, it is stored in the array.    #[inline]
+    pub(crate) fn push(&mut self, value: T) {
+        if self.count < MAX_STACK_CAPACITY {
+            self.initial[self.count] = value;
+            self.count += 1;
+        } else {
+            if self.additional.is_none() {
```

Comment:
It might be more efficient to simply `match` on `self.additional.as_mut()`:

```rust
match self.additional.as_mut() {
    Some(v) => v.push(value),
    None => v = Some(Vec::with_capacity(INITIAL_VEC_CAPACITY)),
}
```

Using `self.additional.is_none()` and `self.additional.as_mut().unwrap()` would lead to more match statements anyway, I guess.

Reply:
Thanks, this looks better. Will update.

Reply:
ok looking again, I think this has to be actually changed as:
```
        match self.additional.as_mut() {
            Some(v) => v.push(value),
            None => {
                let mut vec = Vec::with_capacity(INITIAL_VEC_CAPACITY);
                vec.push(value);
                self.additional = Some(vec);
            }
        }
```

while `match` is more rust idiomatic, in this case, if-else looks easier to follow. We can revisit it if any performance issues are observed.